perun-core/src/main/java/cz/metacentrum/perun/core/implApi/modules/attributes/UserVirtualAttributeCollectedFromUserExtSource.java

```diff
@@ -157,39 +160,66 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
-		patterns.add(setUserExtSourceAttribute);
-		patterns.add(removeUserExtSourceAttribute);
+	@FunctionalInterface
+	public interface AttributeHandleIdenfier {
+		Integer check(AuditEvent ae);
+	}
+
 
-		return patterns;
+	/**
+	 * Returns userId if AuditEvent is correct and null if not.
+	 *
+	 * @return userId
+	 */
+	public List<AttributeHandleIdenfier> getChecks() {
+		List<AttributeHandleIdenfier> list = new ArrayList<>();
+		list.add(ae -> {
+			if (ae instanceof AllAttributesRemovedForUserExtSource) {
+				return ((AllAttributesRemovedForUserExtSource) ae).getUserExtSource().getUserId();
+			} else {
+				return null;
+			}
+		});
+		list.add(ae -> {
+			if (ae instanceof AttributeRemovedForUes && ((AttributeRemovedForUes) ae).getAttribute().getFriendlyName().equals(getSourceAttributeFriendlyName())) {
+				return ((AttributeRemovedForUes) ae).getUes().getUserId();
+			} else {
+				return null;
+			}
+		});
+		list.add(ae -> {
+			if (ae instanceof AttributeSetForUes &&((AttributeSetForUes) ae).getAttribute().getFriendlyName().equals(getSourceAttributeFriendlyName())) {
+				return ((AttributeSetForUes) ae).getUes().getUserId();
+			} else {
+				return null;
+			}
+		});
+		return list;
 	}
 
 	@Override
-	public List<String> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, String message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
-		List<String> resolvingMessages = new ArrayList<>();
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
 		if (message == null) return resolvingMessages;
 
-		if (messageNeedsResolve(message)) {
-			log.debug("Resolving virtual attribute value change for message: " + message);
-			User user = perunSession.getPerunBl().getModulesUtilsBl().getUserFromMessage(perunSession, message);
-			if (user != null) {
-				Attribute attribute = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, getDestinationAttributeName());
-				@SuppressWarnings("unchecked") List<String> attributeValue = (ArrayList<String>) attribute.getValue();
-				String messageAttributeSet;
-				if (attributeValue == null || attributeValue.isEmpty()) {
-					AttributeDefinition attributeDefinition = new AttributeDefinition(attribute);
-					messageAttributeSet = attributeDefinition.serializeToString() + " removed for " + user.serializeToString() + ".";
-				} else {
-					messageAttributeSet = attribute.serializeToString() + " set for " + user.serializeToString() + ".";
+		List<AttributeHandleIdenfier> list = getChecks();
+		for (AttributeHandleIdenfier attributeHandleIdenfier : list) {
+			if (attributeHandleIdenfier.check(message) != null) {
```

Comment:
```suggestion
			Long userId = attributeHandleIdenfier.check(message);
			if (userId != null) {
```
This reduces the need of calling the check method twice.